File: frontend/Gorki/src/app/service/report-service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ReportDTO } from '../model/ui/report-dto';
import { environment } from '../../env/environment';
import { AuthService } from '../infrastructure/auth.service';
import { formatDate } from '@angular/common';

@Injectable({
  providedIn: 'root',
})
export class ReportService {
  constructor(private http:HttpClient, private auth:AuthService){}

  generatePassengerReport(from?:Date | null, to?:Date | null): Observable<ReportDTO>{
    const params: any = {};
    if (from) params.from = formatDate(from, 'yyyy-MM-dd', 'en-US');
    if (to) params.to = formatDate(to, 'yyyy-MM-dd', 'en-US');
    return this.http.get<ReportDTO>(`${environment.apiHost}/users/passenger/${this.auth.getId()}/report`, {params});
  }

  generateDriverReport(from?:Date | null, to?:Date | null): Observable<ReportDTO>{
    const params: any = {};
    if (from) params.from = formatDate(from, 'yyyy-MM-dd', 'en-US');
    if (to) params.to = formatDate(to, 'yyyy-MM-dd', 'en-US');
    return this.http.get<ReportDTO>(`${environment.apiHost}/users/driver/${this.auth.getId()}/report`, {params});
  }

  generateAggregateReport(from?:Date | null, to?:Date | null): Observable<ReportDTO>{
    const params: any = {};
    if (from) params.from = formatDate(from, 'yyyy-MM-dd', 'en-US');
    if (to) params.to = formatDate(to, 'yyyy-MM-dd', 'en-US');
    return this.http.get<ReportDTO>(`${environment.apiHost}/users/all/report`, {params});
  }

  generateAdminUserReport(email:string, from?:Date | null, to?:Date | null): Observable<ReportDTO>{
    const params: any = {};
    params.email = email;
    if (from) params.from = formatDate(from, 'yyyy-MM-dd', 'en-US');
    if (to) params.to = formatDate(to, 'yyyy-MM-dd', 'en-US');
    return this.http.get<ReportDTO>(`${environment.apiHost}/users/report`, {params});
  }
  
}

File: mobile/app/src/main/java/ftn/mrs_team11_gorki/service/ReportService.java
package ftn.mrs_team11_gorki.service;

import ftn.mrs_team11_gorki.dto.ReportDTO;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface ReportService {

    @GET("api/users/passenger/{id}/report")
    Call<ReportDTO> generatePassengerReport(
            @Path("id") Long id,
            @Query("from") String from,   // "yyyy-MM-dd" ili null
            @Query("to") String to        // "yyyy-MM-dd" ili null
    );

    @GET("api/users/driver/{id}/report")
    Call<ReportDTO> generateDriverReport(
            @Path("id") Long id,
            @Query("from") String from,
            @Query("to") String to
    );

    @GET("api/users/all/report")
    Call<ReportDTO> generateAdminAggregateReport(
            @Query("from") String from,
            @Query("to") String to
    );

    @GET("api/users/report")
    Call<ReportDTO> generateAdminUserReport(
            @Query("email") String email,
            @Query("from") String from,
            @Query("to") String to
    );
}
